feat(UX)!: always sharable URLs on List/Report/Kanbanview

ankush opened the pull request:
"Share URL" feature is completely broken:

1. Doesn't show any link right now (copy to clipboard works)
2. If you reshare URL same params are added to URL again. (simple concatenation)
3. Shared URLs aren't working at all, router strips out the search string as first step in routing, so you end up on basic unfiltered list view. 🤡 

https://github.com/frappe/frappe/blob/d666769360ec80b0942dd50a090f5215d37a01e8/frappe/public/js/frappe/router.js#L385-L386



After:

Any change in filters are immediately reflected in URL.

Why? Simple rationale, you can share exactly what you're seeing with someone else without having to know about (_broken_) share URL feature. 

<img width="1286" alt="Screenshot 2022-06-07 at 2 23 26 PM" src="https://user-images.githubusercontent.com/9079960/172339362-6282f845-5213-4eba-b144-c40e6cc2d9e0.png">




- [x] list view refresh URL
- [x] kanban board refresh URL
- [x] Report view 
- [x] refactor - use https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams ?
- [x] UI tests
- [x] saved filters with LIKE filtering



closes https://github.com/frappe/frappe/issues/16040
closes https://github.com/frappe/frappe/issues/16492
closes https://github.com/frappe/frappe/issues/16850 
closes https://github.com/frappe/frappe/issues/16371





`no-docs`

ankush committed:
feat(UX): refresh URL when updating filters

ankush committed:
fix: remove query params from existing URL

ankush committed:
fix: view routing from URL query paramters

ankush committed:
fix: double urlencoding of values

This was breaking URLs for complex filters

ankush committed:
fix: don't forget 'like' filters

Currently like filters get reset if field is part of standard filters

ankush committed:
fix: kanban board from shared links

- Apply filter from URL
- Update URL in browser when filters change

ankush committed:
fix: handle multiselect report options

ankush committed:
feat(ux): Always sharable query report URLs

ankush commented:
@surajshetty3416 @shariquerik @netchampfaris is this acceptable? If so I'll work on further cleanup of code 😬

surajshetty3416 commented:
> @surajshetty3416 @shariquerik @netchampfaris is this acceptable? If so I'll work on further cleanup of the code 😬

Since this is fixing the overall "share listview state" feature and its discoverability, a bit messy URL seems like a decent tradeoff.

LGTM.

surajshetty3416 commented:
> @surajshetty3416 @shariquerik @netchampfaris is this acceptable? If so I'll work on further cleanup of the code 😬

Since this is fixing the overall "share listview state" feature and its discoverability, a bit messy URL seems like a decent tradeoff.

LGTM.

ankush committed:
refactor: update listview URL on all refresh

Previous changes only made it refresh on change of filters, this change
now updates URL ALWAYS.

Co-Auhtored-By: Suraj Shetty <surajshetty3416@gmail.com>

ankush committed:
test: simple test for URL routing with params

TurkerTunali commented:
Thank you @ankush . I was saving my shortcuts to my phone it was real deal breaker for me. Because, I had to re-enter all my filters again and again.
